[Qt] Display transaction fee with sat/vbyte value in SendCoinsDialog

ericallam opened the pull request:
Related to issue #11564, this PR designed to provide feedback to the user about their relative transaction fee before they finish broadcasting their tx to the bitcoin network. 

Displaying the sat/vbyte is useful for knowing how likely their tx will be included in an upcoming block when using public fee estimation tools like https://jochen-hoenicke.de/queue/#24h, https://estimatefee.com, and https://bitcoinfees.earn.com.

This updates the SendCoinsDialog to look like this:

<img width="838" alt="screen shot 2018-01-15 at 12 46 14" src="https://user-images.githubusercontent.com/534/34943625-4f0037a6-f9f4-11e7-8a77-49d35df34c76.png">

ericallam committed:
Display fee sat/vbyte in SendCoinsDialog

Related to issue #11564, this is designed to provide
feedback to the user before they finish broadcasting their tx
to the bitcoin network.

Sjors commented:
I think the transaction confirmation dialog should use the same unit as the send screen. So you could make two pull requests:

1 - add sat / (v)byte to the dropdown on the send screen
2 - add fee per (kilo)byte to the confirmation screen (this PR) 

Nit: I would move the fee per size inside brackets at the end, e.g. "0.00002220 BTC added as transaction fee (0.221 kB at 10.0452 sat/byte)".

ericallam commented:
@Sjors I agree that it's a good idea to do 2 different PRs for these changes. Adding the ability to specify custom fees in sat/vbyte is a decent chunk of work because of the way the Custom Fee text field reuses the `BitcoinAmountField` widget which is used in a couple different places in the UI (like when selecting an amount of BTC to send in a transaction, and to receive). 

I think the best approach for the custom fee section would be to not use `BitcoinAmountField` but to  create a new widget specifically for the custom fee input that would both:

1. Allow inputting the fee as sat/vbyte
2. When not inputting the fee as sat/vbyte, display (either in the UI or as a tooltip) the resulting sat/vbyte.

When this second PR happens, the Recommended section should also probably either show the sat/vbyte instead of BTC/kb, or show both.

I agree with your Nit about the format of the fee per size label, and will be updating this PR to reflect the change in a few minutes.

ericallam commented:
The `SendCoinsDialog` confirmation screen now looks like this:

![screen shot 2018-01-17 at 12 44 17](https://user-images.githubusercontent.com/534/35043357-34f83764-fb84-11e7-92f6-27f413b218b8.png)

jonasschnelli commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -318,11 +318,23 @@ void SendCoinsDialog::on_sendButton_clicked()
         // append transaction size
-        questionString.append(" (" + QString::number((double)currentTransaction.getTransactionSize() / 1000) + " kB)");
+        questionString.append(" (");
+        questionString.append(QString::number(txSize / 1000) + " kB");
```
Use `formatBytes()` from GUIUtils or otherwise please use KB / 1024 for consistency reasons.

jonasschnelli commented:
Concept ACK

ericallam committed:
Using formatBytes() to standardize the size output

PierreRochard commented:
Strong Concept ACK, hopefully we coalesce on a solution to issue #11564

jonasschnelli commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -318,11 +318,24 @@ void SendCoinsDialog::on_sendButton_clicked()
         // append fee string if a fee is required
         questionString.append("<hr /><span style='color:#aa0000;'>");
         questionString.append(BitcoinUnits::formatHtmlWithUnit(model->getOptionsModel()->getDisplayUnit(), txFee));
+
```
Is there a reason for this newline and the double-newline in L333?

ericallam replied:
Fixed in e8485d10e63c1c66ea85b253ff38ee86fd2a3ea6

jonasschnelli commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -318,11 +318,24 @@ void SendCoinsDialog::on_sendButton_clicked()
+        double txSize = (double)currentTransaction.getTransactionSize();
+
+        // safe to assume that txSize cannot be 0 at this point?
+        double satPerVByte = txFee / txSize; // txFee is already in satoshis
```
From the isolated code part perspective, I'd say either factor out into a function that handles the txSize validity or so an if structure above to ensue txSize is valid.

ericallam replied:
Moved sat/vbyte calculation to the WalletModelTransaction class in e8485d10e63c1c66ea85b253ff38ee86fd2a3ea6

jonasschnelli commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -318,11 +318,24 @@ void SendCoinsDialog::on_sendButton_clicked()
         // append transaction size
-        questionString.append(" (" + QString::number((double)currentTransaction.getTransactionSize() / 1000) + " kB)");
+        questionString.append(" (");
+        questionString.append(GUIUtil::formatBytes(txSize));
```
This may loos precision since formatBytes() does show 1.2kb as 1kb... not sure if this is a problem though. Maybe better keep it the way it was.

ericallam replied:
Thanks for the heads up, moved back to original version in e8485d10e63c1c66ea85b253ff38ee86fd2a3ea6

ericallam committed:
Move sat/vbyte calculation to WalletModelTransaction

This allowed isolation of the transaction size validation into the wallet model, to protect against divide by 0 errors.

Also removed use of `GUIUtil::formatBytes` to output the size of the transaction to have better precision in the output.

Sjors reviewed (approved):
tACK e8485d1, but please squash the commits. Looks like @jonasschnelli's items were addressed. 

I'd still like the user to be able to enter a fee in sat/byte, but that can be done on top of this work.

I think we were inclined against using the "v" prefix in #12180, i.e. just `sat/byte` should suffice.

MarcoFalke commented:
Needs rebase if still relevant. (Personally I think we shouldn't expose gui users to the term "vbyte" unless they request raw data about fee estimates)
